ci-operator/step-registry/cucushift/upgrade/toversion/cucushift-upgrade-toversion-commands.sh

```diff
@@ -0,0 +1,487 @@
+            echo "${case_id} is enabled via ENABLE_OTA_TEST on this job."
+            return 0
+        fi
+    done
+    return 1
+}
+
+function set_channel(){
+    x_ver=$( echo "${DUMMY_TARGET_VERSION}" | cut -f1 -d. )
+    y_ver=$( echo "${DUMMY_TARGET_VERSION}" | cut -f2 -d. )
+    ver="${x_ver}.${y_ver}"
+    if ! oc adm upgrade channel candidate-${ver}; then
+        echo "Fail to change channel to candidate-${ver}!"
+        return 1
+    fi
+    local retry=3
+    while (( retry > 0 ));do
+        recommends=$(oc get clusterversion version -o json|jq -r '.status.availableUpdates')
+        if [[ "${recommends}" == "null" ]]; then
+            (( retry -= 1 ))
+            sleep 60
+            echo "No recommended update available! Retry..."
+        else
+            echo "Recommencded update: ${recommends}"
+            break
+        fi
+    done
+    if (( retry == 0 )); then
+        echo "Timeout to get recommended update!" 
+        return 1
+    fi
+}
+
+# judge if upgrade with --to or --to-latest cmd
+# set correct target_version for the upgrade
+function set_target_and_upgrade_cmd(){
+    mapfile -t available_versions < <(echo "$(oc get clusterversion version -o json|jq -r '.status.availableUpdates[]?.version')")
+    if [[ -z "${available_versions[*]}" ]]; then
+        echo "No recommended update available!"
+        exit 1
+    fi
+    # for "oc adm upgrade --to x.y.z" and "oc adm upgrade --to $latest_version_in_available_update"
+    upgrade_cmd="oc adm upgrade --to ${TARGET_VERSION}"
+    # x.y.z: run upgrade with "oc adm upgrade --to x.y.z";
+    if [[ -n "${UPGRAGE_TO_VERSION}" ]] && [[ "${UPGRAGE_TO_VERSION}" != "latest" ]]; then
+        valid_target="false"
+        # check if x.y.z is in available_versions, if not, it's invalid target with --to.
+        for version in ${available_versions[*]}; do
+            if [[ "${UPGRAGE_TO_VERSION}" == "${version}" ]]; then
+                TARGET_VERSION="${UPGRAGE_TO_VERSION}"
+                valid_target="true"
+                break
+            fi
+        done
+        if [[ "${valid_target}" == "false" ]]; then
+            echo "Specified target version ${UPGRAGE_TO_VERSION} is not in recommended update, upgrade will not continue!"
+            exit 1
+        fi
+    # get target_version from available updates
+    else
+        TARGET_VERSION=$(oc get clusterversion version -o json|jq -r '.status.availableUpdates[]?.version'|sort|tail -n1)
+        # there will be three types version in candidate channel:4.15.0-ec1,4.15.0-rc1,4.15.0
+        # when there are all above, sort will return 4.15.0-rc1 as latest(unexpected, should be 4.15.0),
+        # when there is both 4.15.0 and 4.15.0-rc1, sort will return 4.15.0-rc1 as latest(unexpected, should be 4.15.0),
+        # when there is both 4.15.0-rc1 and 4.15.0-ec1, sort will return 4.15.0-rc1 as latest(expected),
+        # there will not be 4.15.0 and 4.15.0-ec1 only, ignore
```

Comment:
If there is 4.15.0, 4.16-ec.1 in `availableUpdates`, the logic will get wrong results.

Reply:
4.16 should always > 4.15 in sort, right?